# Theory/Computation Problems

### Problem 1 (20 points) 
Show that the stationary point (zero gradient) of the function
$$
\begin{aligned}
    f=2x_{1}^{2} - 4x_1 x_2+ 1.5x^{2}_{2}+ x_2
\end{aligned}
$$
is a saddle (with indefinite Hessian). Find the directions of downslopes away from the saddle. Hint: Use Taylor's expansion at the saddle point. Find directions that reduce $f$.

#### Solution 

$$f(x)=2x_1^2 - 4x_1 x_2+ 1.5x^2_2+ x_2$$

The gradient of the function $f(x)$ is: 

$$g(x)=\nabla f(x)=
    \begin{bmatrix}
    \frac{\partial f}{\partial x_1}\\\\
    \frac{\partial f}{\partial x_2}
    \end{bmatrix}
    =
    \begin{bmatrix}
    4x_1-4x_2\\\\
    -4x_1+3x_2+1
    \end{bmatrix}
$$

Let $x^\star$ be a stationary point of $f(x)$. Then, $x^\star$ satisfies the
equation $g(x^\star)=0$.

$$\Rightarrow 
    g(x^\star)=
    \begin{bmatrix}
    4x_1-4x_2\\\\
    -4x_1+3x_2+1
    \end{bmatrix}
    =
    \begin{bmatrix}
    0\\\\
    0
    \end{bmatrix}
$$

<br>
$$\Rightarrow 
  \begin{equation*}
  \left\{\begin{aligned}
    4x_1-4x_2=0\\
    -4x_1+3x_2+1=0
  \end{aligned}
  \right.\end{equation*} 
$$

<br>
$$\Rightarrow 
  \begin{equation*}
  \left\{\begin{aligned}
    x_1 &=x_2\\
    -4x_1+3x_1+1 &=0
  \end{aligned}
  \right.\end{equation*} 
$$

<br>
$$\begin{aligned}
  \Rightarrow x_1=1=x_2\\
  \Rightarrow x^\star=
  \begin{bmatrix}
    1\\
    1
  \end{bmatrix}\end{aligned}
$$


<br> 
The Hessian matrix of $f(x)$ is:

$$\begin{aligned}
  H(x)=\nabla ^2f(x)=
  \begin{bmatrix}
    \frac{\partial ^2 f}{\partial x_1^2} && \frac{\partial ^2 f}{\partial x_1\partial x_2}\\\\
    \frac{\partial ^2 f}{\partial x_2\partial x_1} && \frac{\partial ^2 f}{\partial x_2^2}
  \end{bmatrix}
  =
  \begin{bmatrix}
    4 && -4\\\\
    -4 && 3
  \end{bmatrix}\end{aligned}
$$

and $H(x^\star)= 
  \begin{bmatrix}
    4 && -4\\\\
    -4 && 3
  \end{bmatrix}$.

<br> 
We have that: <br> 
$$|H(x)|=\lambda_1 \cdot \lambda_2$$ 
and also: <br> 
 $$\begin{aligned}
    |H(x)|=3\cdot 4-(-4)\cdot (-4)=12-16=-4<0
    \end{aligned}
 $$

$\Rightarrow \lambda_1,\lambda_2$ have opposite signs. Hence, the matrix
$H(x)$ is indefinite and therefore, the stationary point is a **saddle point**. 


<br />
The 2nd-order Taylor Series approximation of the function is:
$$f(x)\approx f(x^\star)+g^T(x^\star)(x-x^\star)+\frac{1}{2}(x-x^\star)^TH(x^\star)(x-x^\star)$$

where 

$f(x^\star)=2\cdot 1^2-4\cdot 1\cdot 1+1.5\cdot 1^2+1=0.5$
<br>
$g(x^\star)=0$

$$\Rightarrow f(x)\approx 0.5+\frac{1}{2} 
  \begin{bmatrix}
    x_1-1 && x_2-1
  \end{bmatrix} 
  \begin{bmatrix}
    4 && -4\\\\
    -4 && 3
  \end{bmatrix}
  \begin{bmatrix}
    x_1-1 \\\\
    x_2-1
  \end{bmatrix}
  =2x_1^2 - 4x_1 x_2+ 1.5x^2_2+ x_2
$$


At the saddle point, in the directions that $f$ reduces, we have that 
$f(x)< f(x^\star)$.

$$\Rightarrow f(x)-f(x^\star)< 0 $$
$$\Rightarrow \frac{1}{2}(x-x^\star)^TH(x^\star)(x-x^\star)< 0 $$
$$\Rightarrow (x-x^\star)^TH(x^\star)(x-x^\star)< 0 $$
$$\Rightarrow 4x_1^2 - 8x_1 x_2+ 3x^2_2+ 2x_2 -1< 0 $$
$$\Rightarrow (2x_1-3x_2+1)(2x_1-x_2-1)< 0 $$


which means that $f$ reduces when:

$$(2x_1-3x_2+1)>0  \ and\ (2x_1-x_2-1)< 0 $$
or
$$(2x_1-3x_2+1)< 0 \ and\ (2x_1-x_2-1)> 0 $$



More specifically, the eigenvalues of the Hessian are:


In [ ]:
import numpy as np
from numpy import linalg as LA

H = np.array([[4, -4], [-4, 3]])  # Hessian
w,v = LA.eig(H)                   # Eigenvalues and eigenvectors 
print("The eigenvalues are: {}".format(w))
print("The eigenvectors are:\n {}".format(v))

The eigenvalues are: [ 7.53112887 -0.53112887]
The eigenvectors are:
 [[ 0.74967818  0.66180256]
 [-0.66180256  0.74967818]]


$\lambda_1=7.53112887, \lambda_2=-0.53112887$

The direction of downslope, where the function curves down, is the principal direction which corresponds to the negative eigenvalue. This is:

$$v_2=\begin{bmatrix}
    -0.66180256\\\\
    0.74967818
  \end{bmatrix}$$

### Problem 2 (50 points) 

* (10 points) Find the point in the plane $x_1+2x_2+3x_3=1$ in $\mathbb{R}^3$ that is nearest to the point $(-1,0,1)^T$. Is this a convex problem? Hint: Convert the problem into an unconstrained problem using $x_1+2x_2+3x_3=1$.

* (40 points) Implement the gradient descent and Newton's algorithm for solving the problem. Attach your codes along with a short summary including (1) the initial points tested, (2) corresponding solutions, (3) a log-linear convergence plot.

#### Solution

i) If we let 
$w=\begin{bmatrix}
    x_1 & x_2 & x_3
  \end{bmatrix}^T$
be any point on the plane and 
$x_0=\begin{bmatrix}
    -1 & 0 & 1
  \end{bmatrix}^T$
be the given point of interest, then the vector

$$z=\begin{bmatrix}
    x_1-(-1)\\\\ 
    x_2-0\\\\
    x_3-1
  \end{bmatrix}
  =
  \begin{bmatrix}
    x_1+1\\\\ 
    x_2\\\\
    x_3-1
  \end{bmatrix}
$$

denotes the distance between them. The point $w$ that is nearest to $x_0$ is the one with the smallest distance, i.e. whose vector $z$ length is minimum.
The length of the vector $z$ is:
$$|z|=\sqrt{(x_1+1)^2+(x_2)^2+(x_3-1)^2}$$
<br>
Therefore, we need to find the solution to the following minimization problem: 

$$
\begin{aligned}
&\underset{x_1,x_2,x_3}{\text{minimize:}} && (x_1+1)^2+(x_2)^2+(x_3-1)^2 \\\\
&\text{subject to:} && x_1+2x_2+3x_3=1 \\\\
&&& x_i \in \Re, \ i=1,2,3
\end{aligned}
$$

<br>
If we let $x_1=-2x_2-3x_3+1$, we have that:
$$
(x_1+1)^2+(x_2)^2+(x_3-1)^2 = (2-2x_2-3x_3)^2+(x_2)^2+(x_3-1)^2
$$

and we can convert the problem to the following unconstrained optimization problem:

$$
\underset{x_2,x_3}{\text{minimize: }}  f(x)=(2-2x_2-3x_3)^2+(x_2)^2+(x_3-1)^2 
$$

The solution to the minimization problem is:

In [8]:
from scipy.optimize import minimize

# Objective function
fun = lambda x: (2-2*x[0]-3*x[1])**2 + (x[0])**2 + (x[1]-1)**2

# Initial guess
x0 = (1, 0)

# Optimization Result
res = minimize(fun, x0, method='SLSQP')

# Function value
print('Function value: f = {}'.format(res.fun))

# Variable values 
print('Variable values:\nx2, x3 = {}, {}'.format(res.x[0], res.x[1]))

# Compute x1
x1=-2*res.x[0]-3*res.x[1]+1
print("x1 = {}".format(x1))

Function value: f = 0.07142857142857148
Variable values:
x2, x3 = -0.14285713906428532, 0.7857142823269022
x1 = -1.071428568852136


Therefore, the point in the plane that is nearest to $x_0$ is: 
$w=\begin{bmatrix}
    x_1 & x_2 & x_3
  \end{bmatrix}^T
  =
  \begin{bmatrix}
    -1.071428 & -0.142857 & 0.785714
  \end{bmatrix}^T $

The gradient of $f(x)$ is:
$$g(x)=\nabla f(x)=
    \begin{bmatrix}
    10x_2+12x_3-8\\\\
    12x_2+20x_3-14
    \end{bmatrix}
$$

The Hessian is:
$$\begin{aligned}
  H(x)=\nabla ^2f(x)=
  \begin{bmatrix}
    10 && 12\\\\
    12 && 20
  \end{bmatrix}\end{aligned}
$$


In [9]:
import numpy as np
from numpy import linalg as LA

H = np.array([[10, 12], [12, 20]])  # Hessian
w,v = LA.eig(H)                     # Eigenvalues and eigenvectors 
print("The eigenvalues are: {}".format(w))

The eigenvalues are: [ 2. 28.]


The eigenvalues are all positive, hence the Hessian is positive definite everywhere. Therefore, $f(x)$ and thus the unconstrained optimization problem are convex.

<br>
ii)



### Problem 3 (10 points) 
Let $f(x)$ and $g(x)$ be two convex functions defined on the convex set $\mathcal{X}$. 
* (5 points) Prove that $af(x)+bg(x)$ is convex for $a>0$ and $b>0$. 
* (5 points) In what conditions will $f(g(x))$ be convex?

### Problem 4 (bonus 10 points)
Show that $f({\bf x}_1) \geq f(\textbf{x}_0) + 
    \textbf{g}_{\textbf{x}_0}^T(\textbf{x}_1-\textbf{x}_0)$ for a convex function $f(\textbf{x}): \mathcal{X} \rightarrow \mathbb{R}$ and for $\textbf{x}_0$, $\textbf{x}_1 \in \mathcal{X} 
$. 

# Design Problems

### Problem 5 (20 points) 
Consider an illumination problem: There are $n$ lamps and $m$ mirrors fixed to the ground. The target reflection intensity level is $I_t$. The actual reflection intensity level on the $k$th mirror can be computed as $\textbf{a}_k^T \textbf{p}$, where $\textbf{a}_k$ is given by the distances between all lamps to the mirror, and $\textbf{p}:=[p_1,...,p_n]^T$ are the power output of the lamps. The objective is to keep the actual intensity levels as close to the target as possible by tuning the power output $\textbf{p}$.

* (5 points) Formulate this problem as an optimization problem. 
* (5 points) Is your problem convex?
* (5 points) If we require the overall power output of any of the $n$ lamps to be less than $p^*$, will the problem have a unique solution?
* (5 points) If we require no more than half of the lamps to be switched on, will the problem have a unique solution?

# Note

For this homework, you may want to attach sketches as means to explain your ideas. Here is how you can attach images.

![everly1](img/everly7.jpg)